### PR TITLE
NEW PROVIDER: AkamaiEdgeDNS

### PR DESCRIPTION
**Pull request by svernick**
This PR introduces a provider for Akamai EdgeDNS.

This PR also includes a change to downcase TLSA data.  TLSA data is a hexadecimal string and is case insensitive.  See pull request 1171.

### CODE REVIEW

File: integrationTest/providers.json [hunk 1]
```diff
@@ -58,6 +58,15 @@
     "api_key": "$DNSMADEEASY_API_KEY",
     "secret_key": "$DNSMADEEASY_SECRET_KEY"
   },
+  "EDGEDNS": {
```
Comment (tlimoncelli) [inline integrationTest/providers.json:61]: Would you be open to a name that includes Akamai?  Outsiders are more familiar with the name Akamai than EdgeDNS.  

We learned the hard way that these names are difficult to change.  As a result we have "GCLOUD" and "oracle" and "route53" which aren't optimal names.  I'd like to make sure we put a lot of thought into any new names.

File: pkg/js/parse_tests/015-tlsa/foo.com.zone [hunk 2]
```diff
@@ -1,2 +1,2 @@
 $TTL 300
-_443._tcp        IN TLSA  3 1 1 MDFiYTQ3MTljODBiNmZlOTExYjA5MWE3YzA1MTI0YjY0ZWVlY2U5NjRlMDljMDU4ZWY4Zjk4MDVkYWNhNTQ2YiAgLQo=
+_443._tcp        IN TLSA  3 1 1 mdfiytq3mtljodbinmzlotexyja5mwe3yza1mti0yjy0zwvly2u5njrlmdljmdu4zwy4zjk4mdvkywnhntq2yiaglqo=
```
Comment (tlimoncelli) [inline pkg/js/parse_tests/015-tlsa/foo.com.zone:2]: I'm surprised that this needed to change for the tests to pass.  These tests are agnostic to the providers.  Was this change necessary?

File: providers/edgedns/edgeDnsProvider.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+	api := &edgeDNSProvider{
+		contractID: contractID,
+		groupID:    groupID,
+	}
+	return api, nil
+}
+
+// AuditRecords returns an error if any records are not supportable by this provider.
+func AuditRecords(records []*models.RecordConfig) error {
```
Comment (tlimoncelli) [inline providers/edgedns/edgeDnsProvider.go:97]: Please move the function `AuditRecords to a file called `auditrecords.go`

File: providers/edgedns/edgeDnsService.go [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+import (
+	"fmt"
+	"github.com/StackExchange/dnscontrol/v3/models"
+	"github.com/StackExchange/dnscontrol/v3/pkg/printer"
+	dnsv2 "github.com/akamai/AkamaiOPEN-edgegrid-golang/configdns-v2"
+	"github.com/akamai/AkamaiOPEN-edgegrid-golang/edgegrid"
+)
+
+// EDInitialize initializes the "Akamai OPEN EdgeGrid" library
+func EDInitialize(clientSecret string, host string, accessToken string, clientToken string) {
```
Comment (tlimoncelli) [inline providers/edgedns/edgeDnsService.go:20]: Do the the `ED*` functions need to be exported?  They might... I'm just not seeing why right off the bat.  You might try changing one and see if go-lint or go-vet complain.

If they don't need to be exported, consider calling the functions initialize, zoneDoesExist, createZone, etc.

File: providers/edgedns/edgeDnsService.go [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+}
+
+// EDCreateZone create a new zone and creates SOA and NS records for the zone.
+// Akamai assigns a unique set of authoritative nameservers for each contract. These authorities should be
+// used as the NS records on all zones belonging to this contract.
+func EDCreateZone(zonename string, contractID string, groupID string) error {
+	zone := &dnsv2.ZoneCreate{
+		Zone:                  zonename,
+		Type:                  "PRIMARY",
+		Comment:               "This zone created by DNSControl (https://stackexchange.github.io/dnscontrol/)",
```
Comment (tlimoncelli) [inline providers/edgedns/edgeDnsService.go:46]: Please use the url `http://dnscontrol.org` instead.

**Review by tlimoncelli — COMMENTED** [review]
It's awesome to see a new provider, especially coming from the company themselves!

Everything looks good so far.  I have one big request about the name.  See below.

File: integrationTest/providers.json (hunk 1 above)
Comment (svernick) [reply integrationTest/providers.json]: Done.  Now "akamaiedgedns".

File: pkg/js/parse_tests/015-tlsa/foo.com.zone (hunk 2 above)
Comment (svernick) [reply pkg/js/parse_tests/015-tlsa/foo.com.zone]: You are absolutely right.  I swear that without that change, unit tests failed.  It just proves that I shouldn't swear.  :)
I changed it back to the original.

File: providers/edgedns/edgeDnsProvider.go (hunk 3 above)
Comment (svernick) [reply providers/edgedns/edgeDnsProvider.go]: Done.

File: providers/edgedns/edgeDnsService.go (hunk 4 above)
Comment (svernick) [reply providers/edgedns/edgeDnsService.go]: Done.  Thanks for catching that!

File: providers/edgedns/edgeDnsService.go (hunk 5 above)
Comment (svernick) [reply providers/edgedns/edgeDnsService.go]: Done.

File: pkg/js/parse_tests/015-tlsa/foo.com.zone (hunk 2 above)
Comment (svernick) [reply pkg/js/parse_tests/015-tlsa/foo.com.zone:2]: The automated build failed like this:
--- FAIL: TestParsedFiles (0.88s)
18
    --- FAIL: TestParsedFiles/015-tlsa.js (0.02s)
19
        js_test.go:117: 
20
            	Error Trace:	js_test.go:117
21
            	Error:      	Not equal: 
22
            	            	expected: "$TTL 300\n_443._tcp        IN TLSA  3 1 1 MDFiYTQ3MTljODBiNmZlOTExYjA5MWE3YzA1MTI0YjY0ZWVlY2U5NjRlMDljMDU4ZWY4Zjk4MDVkYWNhNTQ2YiAgLQo=\n"
23
            	            	actual  : "$TTL 300\n_443._tcp        IN TLSA  3 1 1 mdfiytq3mtljodbinmzlotexyja5mwe3yza1mti0yjy0zwvly2u5njrlmdljmdu4zwy4zjk4mdvkywnhntq2yiaglqo=\n"

It appeared to pass on my local system because I had run the test successfully before and the test had been "cached".

To clear the cached tests:
 go clean -testcache

After that, the unit test failed on my local machine.

pkg/js/parse_tests/015-tlsa/foo.com.zone once again contains lowercase.  The test passes.

**Comment by svernick** [comment]
I forgot to run 'go generate' before the previous commit.  My apologies.

**Comment by svernick** [comment]
Hi.  Is there anything that I need to do before this PR can be merged?  Thanks!

File: docs/_providers/akamaiedgedns.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
+---
+name: AkamaiEdgeDns
+title: Akamai Edge DNS Provider
+layout: default
+jsId: AKAMAIEDGEDNS
+---
+# Akamai Edge DNS Provider
```
Comment (tlimoncelli) [inline docs/_providers/akamaiedgedns.md:7]: Please leave a blank line before and after any `#` and before and after any code fence (three ` in a row)

File: docs/_providers/akamaiedgedns.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    "access_token": "akaa-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
+    "client_token": "akaa-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
+    "contract_id": "X-XXXX",
+    "group_id": "NNNNNN"
+}
+{% endhighlight %}
+
+## Usage
+A new zone created by DNSControl:
+```
```
Comment (tlimoncelli) [inline docs/_providers/akamaiedgedns.md:38]: add a blank line before this and others like it.

**Review by tlimoncelli — COMMENTED** [review]
Hi!  Sorry for the delay.  It's been quite a week!

This is nearly ready.  I made a few inline comments.  One additional thing:

* `docs/provider-list.md`  Add AKAMAIEDGEDNS along with your git name.

I think that's it!

File: docs/_providers/akamaiedgedns.md (hunk 6 above)
Comment (svernick) [reply docs/_providers/akamaiedgedns.md:8]: Done

File: docs/_providers/akamaiedgedns.md (hunk 7 above)
Comment (svernick) [reply docs/_providers/akamaiedgedns.md:43]: Done

**Comment by svernick** [comment]
I _think_ that I made all of the required corrections.  If not, I'm happy to do so.
Thanks very much!

**Review by tlimoncelli — APPROVED** [review]

**Comment by tlimoncelli** [comment]
Thank you for contributing this new provider!  I'm excited to have Akamai join the DNSControl family!

**Comment by svernick** [comment]
Tom,

Thank you so much for your expertise and direction!

DNSControl is quite impressive.  The documentation is outstanding.  The code is outstanding.  The test framework is outstanding.

It was a pleasure to code the provider.

Akamai is excited to offer DNSControl to our customers.

Thanks again!

/Steve


From: Tom Limoncelli ***@***.***>
Reply-To: StackExchange/dnscontrol ***@***.***>
Date: Tuesday, June 22, 2021 at 10:25 AM
To: StackExchange/dnscontrol ***@***.***>
Cc: "Vernick, Steve" ***@***.***>, Author ***@***.***>
Subject: Re: [StackExchange/dnscontrol] NEW PROVIDER: AkamaiEdgeDNS (#1174)


Thank you for contributing this new provider! I'm excited to have Akamai join the DNSControl family!
